boehm_gc: bump version to 7.6.4

ivmai opened the pull request.

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,82 @@
+HOMEPAGE="http://www.hboehm.info/gc/"
+COPYRIGHT="1988, 1989 Hans-J. Boehm, Alan J. Demers
+	1991-1996 Xerox Corporation.
+	1996-1999 Silicon Graphics.
```
```diff
 COPYRIGHT="1988, 1989 Hans-J. Boehm, Alan J. Demers
-	1991-1996 Xerox Corporation.
+	1991-1996 Xerox Corporation
-	1996-1999 Silicon Graphics.
+	1996-1999 Silicon Graphics
+	1999-2011 Hewlett-Packard Development Company, L.P."
```

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,82 @@
+	1996-1999 Silicon Graphics.
+	1999-2011 Hewlett-Packard Development Company, L.P."
+LICENSE="BOEHM"
+REVISION="2"
```
```diff
-LICENSE="BOEHM"
+LICENSE="Boehm"
-REVISION="2"
+REVISION="1"
```
You also need to "attach" a copy of the `Boehm` (license) file to this recipe (in `haikuports/dev-libs/boehm_gc/licenses/`).
(That's not required by very popular licenses, but this one is not yet "available" in a default Haiku install.)

ivmai replied:
Hmm. BOEHM is already used by 7.6.0 recipe. Should I rename it (to Boehm) in both recipes?

fbrosson replied:
Sorry, I had not noticed the file was already there.
I guess it would be OK to keep it as `BOEHM`.

But maybe it would be better to change it to `Boehm` and rename the file.
Let's see what the other contributors think about this.
(I'd say that this is not an acronym, so it should probably be changed to `Boehm`, but I'm OK with keeping it as is.)

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,82 @@
+PATCHES="boehm_gc-$portVersion.patchset"
+
+ARCHITECTURES="?x86 ?x86_gcc2 x86_64"
+SECONDARY_ARCHITECTURES="?x86 ?x86_gcc2"
```
```diff
-ARCHITECTURES="?x86 ?x86_gcc2 x86_64"
+ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
```
(We like to list `x86_gcc2`, `x86` and `x86_64` using this order.)
Some key contributors also don't mention `x86_gcc2` in `SECONDARY_ARCHITECTURES` anymore, but that's OK for me to have it. This is because the only "flavor" of Haiki which has x86_gcc2 as a secondary arch is x86(gcc5)(hybrid), which is not supported anymore (and is therefore "stuck" at `hrev51205`).

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,82 @@
+				 --enable-thread-local-alloc \
+				 --enable-parallel-mark \
+				 --disable-debug \
+				 --disable-dependency-tracking
```
```diff
 	runConfigure ./configure --enable-large-config \
-				 --enable-threads=posix \
+		 --enable-threads=posix \
-				 --enable-thread-local-alloc \
+		 --enable-thread-local-alloc \
-				 --enable-parallel-mark \
+		 --enable-parallel-mark \
-				 --disable-debug \
+		 --disable-debug \
-				 --disable-dependency-tracking
+		 --disable-dependency-tracking
```

extrowerk commented:
Thanks, nice Job!
Have you tested it on 64 bit?

fbrosson commented:
The only recipes which depend on this one are:
* `www-client/w3m/w3m-0.5.3.recipe`
* `dev-lisp/ecl/ecl-16.1.3.recipe`

So I guess it would be nice to drop `boehm_gc-7.6.0.recipe`, since you are adding `boehm_gc-7.6.4.recipe`.
And before the merge we can try to build `w3m` and `ecl` (against the libs provided by `boehm_gc-7.6.4`) to make sure there are no regressions.

extrowerk commented:
The earlier Boehm compiles just fine on x86_64, but it fails during the Test().
I was unable to fix it, but according to a short look at the patchset, ivmai using an unchanged patchset. Ivmai, please coorect me if i wrong.

So, we do need to TEST() it.

ivmai commented:
I've done all mentioned cosmetic edits.

> Have you tested it on 64 bit?

Not yet.

> ivmai using an unchanged patchset. 

Almost unchanged, so it should work the same as for 7.6.0.

ivmai commented:
The commit (https://github.com/ivmai/bdwgc/commit/847927f0e71ef1a8e3dcf2ee0879bf416311daae) in upstream master is slightly different from boehm_gc-7.6.4.patchset in this repository

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+INSTALL()
+{
+	make install
+	prepareInstalledDevelLibs libgc libcord
```
```diff
 	make install
+	rm -f "$libDir"/lib*.la
 	prepareInstalledDevelLibs libgc libcord
```
(We've decided to drop the libtool files 1 or 2 years ago.)
Thanks!

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+
+TEST()
+{
+	make check -e CFLAGS+="-fPIC"
```
It builds fine on x86_64 and x86 (secondary arch), but when I run `haikuporter --test boehm_gc` (on x86_64) or `haikuporter --test boehm_gc_x86` (on x86_gcc2) I hit this error:
* on x86_64:
```
libtool: link: gcc -fexceptions -fPIC -o .libs/threadleaktest tests/thread_leak_test.o  ./.libs/libgc.so -L/packages/libatomic_ops-7.6.2-1/.self/develop/lib -latomic_ops -lpthread -Wl,-rpath -Wl,/packages/boehm_gc-7.6.4-1/.self/lib
tests/thread_leak_test.o: In function `main':
thread_leak_test.c:(.text+0xca): undefined reference to `GC_pthread_create'
thread_leak_test.c:(.text+0x128): undefined reference to `GC_pthread_join'
collect2: error: ld returned 1 exit status
```
* on x86 2ndary arch:
```
libtool: link: gcc -fexceptions -fPIC -o .libs/threadleaktest tests/thread_leak_test.o  ./.libs/libgc.so -L/packages/libatomic_ops_x86-7.6.2-1/.self/develop/lib/x86 -latomic_ops -lpthread -Wl,-rpath -Wl,/packages/boehm_gc_x86-7.6.4-1/.self/lib/x86
tests/thread_leak_test.o: In function `main':
thread_leak_test.c:(.text+0xd1): undefined reference to `GC_pthread_create'
thread_leak_test.c:(.text+0x129): undefined reference to `GC_pthread_join'
collect2: error: ld returned 1 exit status
```
I did not investigate yet.

alaviss replied:
I believe having `CFLAGS+="-fPIC"` overriden the `CFLAGS` used during build. I've not tested but may be this would fix the issue:

```diff
-	make check -e CFLAGS+="-fPIC"
+	make check
```

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
+
+portVersionCompat="$portVersion compat >= 7"
```
```diff
 ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
-SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
+SECONDARY_ARCHITECTURES="x86"
 
+libcordVersion="1.3.0"
+libgcVersion="1.3.2"
+libcordVersionCompat="$libcordVersion compat >= ${libcordVersion%%.*}"
+libgcVersionCompat="$libgcVersion compat >= ${libgcVersion%%.*}"
-portVersionCompat="$portVersion compat >= 7"
+portVersionCompat="$portVersion compat >= ${portVersion%%.*}"

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+PROVIDES="
+	boehm_gc$secondaryArchSuffix = $portVersionCompat
+	lib:libgc$secondaryArchSuffix = $portVersionCompat
+	lib:libcord$secondaryArchSuffix = $portVersionCompat
```
```diff
 PROVIDES="
 	boehm_gc$secondaryArchSuffix = $portVersionCompat
-	lib:libgc$secondaryArchSuffix = $portVersionCompat
-	lib:libcord$secondaryArchSuffix = $portVersionCompat
+	lib:libcord$secondaryArchSuffix = $libcordVersionCompat
+	lib:libgc$secondaryArchSuffix = $libgcVersionCompat
 	"
```
(`$portVersion` is only OK for the first line. For the `lib:` and `devel:` lines we have different values.)
(BTW, we like to sort alphabetically the `lib:` lines.)

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	lib:libatomic_ops${secondaryArchSuffix}
```
```diff
-	lib:libatomic_ops${secondaryArchSuffix}
+	lib:libatomic_ops$secondaryArchSuffix
```
(No need for the curly braces here.)

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+PROVIDES_devel="
+	boehm_gc${secondaryArchSuffix}_devel = $portVersionCompat
+	devel:libgc${secondaryArchSuffix} = $portVersionCompat
+	devel:libcord${secondaryArchSuffix} = $portVersionCompat
```
```diff
 PROVIDES_devel="
 	boehm_gc${secondaryArchSuffix}_devel = $portVersionCompat
-	devel:libgc${secondaryArchSuffix} = $portVersionCompat
-	devel:libcord${secondaryArchSuffix} = $portVersionCompat
+	devel:libcord$secondaryArchSuffix = $libcordVersionCompat
+	devel:libgc$secondaryArchSuffix = $libgcVersionCompat
 	"
```

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+	cmd:autoconf
+	cmd:make
+	cmd:gcc$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
```
```diff
 BUILD_PREREQUIRES="
 	haiku${secondaryArchSuffix}_devel
-	cmd:libtoolize$secondaryArchSuffix
 	cmd:autoconf
-	cmd:make
 	cmd:gcc$secondaryArchSuffix
 	cmd:ld$secondaryArchSuffix
+	cmd:libtoolize$secondaryArchSuffix
+	cmd:make
 	cmd:pkg_config$secondaryArchSuffix
 	"
+
+defineDebugInfoPackage boehm_gc$secondaryArchSuffix \
+	"$libDir"/libcord.so.$libcordVersion \
+	"$libDir"/libgc.so.$libgcVersion
+
 BUILD()
```

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,81 @@
+			--enable-threads=posix \
+			--enable-parallel-mark \
+			--disable-debug \
+			--disable-dependency-tracking
```
```diff
	runConfigure ./configure --enable-large-config \
-			--enable-threads=posix \
+		--enable-threads=posix \
-			--enable-parallel-mark \
+		--enable-parallel-mark \
-			--disable-debug \
+		--disable-debug \
-			--disable-dependency-tracking
+		--disable-dependency-tracking
```
(One tab removed in 4 lines.)

fbrosson commented:
FWIW, 7.6.**0** passes 13 tests out of 14 on x86 secondary arch:
```
PASS: cordtest
PASS: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
PASS: threadkey_test
FAIL: subthreadcreate_test
PASS: initsecondarythread_test
PASS: disclaim_test
PASS: disclaim_bench
============================================================================
Testsuite summary for gc 7.6.0
============================================================================
# TOTAL: 14
# PASS:  13
# FAIL:  1
```
And on x86_64, 7.6.**0** there is a crash and the the test suite hangs:
```
PASS: cordtest
./test-driver: line 107: 75373 Abort                   "$@" > $log_file 2>&1
FAIL: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
```
The system then becomes unstable. But this might be due to a few tweaks I did to my Haiku x86_64 for a reason that would be to long to explain.

ivmai commented:
All fixed except for "we need to mark it as untested on all architectures".

ivmai commented:
According to the test suite logs above, 7.6.0 also has test fails, so no regression.

ivmai commented:
> thread_leak_test.c:(.text+0xd1): undefined reference to `GC_pthread_create'

I know how to fix it.

alaviss commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,90 @@
+	1999-2011 Hewlett-Packard Development Company, L.P."
+LICENSE="Boehm"
+REVISION="1"
+SOURCE_URI="https://github.com/ivmai/bdwgc/releases/download/v$portVersion/gc-$portVersion.tar.gz"
```
Shorter link

```diff
-SOURCE_URI="https://github.com/ivmai/bdwgc/releases/download/v$portVersion/gc-$portVersion.tar.gz"
+SOURCE_URI="http://www.hboehm.info/gc/gc_source/gc-$portVersion.tar.gz"

fbrosson replied:
It's shorter, but it's not https and the first URI where the archive was available from is on GitHub, as we can see with the timestamps:
```
TZ=UTC ls -go --fu github/gc-7.6.4.tar.gz shorter-uri/gc-7.6.4.tar.gz
-rw-r--r-- 1 1120784 2018-01-26 13:19:23.000000000 +0000 github/gc-7.6.4.tar.gz
-rw-r--r-- 1 1120784 2018-01-29 03:06:48.000000000 +0000 shorter-uri/gc-7.6.4.tar.gz
```
Moreover, on GutHub we are sure that the archives won't be removed, whereas on bboehm.info old tarballs might be removed.

alaviss commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,90 @@
+libcordVersion="1.3.0"
+libgcVersion="1.3.2"
+libcordVersionCompat="$libcordVersion compat >= ${libcordVersion%%.*}"
+libgcVersionCompat="$libgcVersion compat >= ${libgcVersion%%.*}"
```
It's prefered that one uses camelCase for local variables, but it's not strictly required.

```diff
-libcordVersion="1.3.0"
-libgcVersion="1.3.2"
-libcordVersionCompat="$libcordVersion compat >= ${libcordVersion%%.*}"
-libgcVersionCompat="$libgcVersion compat >= ${libgcVersion%%.*}"
+libVersion="1.3.2"
+libCordVersion="1.3.0"
+libVersionCompat="$libVersion compat >= ${libVersion%%.*}"
+libCordVersionCompat="$libCordVersion compat >= ${libCordVersion%%.*}"

fbrosson replied:
Well, if we consider that libcord and libgc are words, then we already have a sort of camelCase :)

fbrosson replied:
BTW, keeping the basename of the libraries is nice because with recipes that have many libraries we can use the eval trick I recently used for ffmpeg.
And the wiki does not mention camelCase. `libVersion` and ``libVersionCompat` happen to be camelCase, but when we have more than just one library, then I think it is nice to use the original basename :)

ivmai commented:
> You appears to be the maintainer of boehm. If that's the case, would you be interested in upstreaming Haiku patches?

Yes, the patch enabling boehm-gc on Haiku is already there (slightly different version, I will check the difference soon) - see ivmai/bdwgc@847927f
And, I'm ready to accept other patches (to master branch), just send a pull request.

> Shorter link
> SOURCE_URI="http://www.hboehm.info/...

I haven't changed this because Travis CI puts the tarball to https://github.com/ivmai/bdwgc/releases/... first and later the tarball is copied to http://www.hboehm.info/... manually.

> make check -e CFLAGS+="-fPIC"

I've removed -e ...

Also, I've removed --enable-threads=posix (because it is on by default) and --disable-debug (I don't see a reason to remove some boehm-gc debugging support API).

ivmai commented:
> thread_leak_test.c:(.text+0xd1): undefined reference to `GC_pthread_create'

To my understanding, CFLAGS+="-fPIC" was the only reason of the undefined reference error (as well as of threaded tests failures), @fbrosson could you please recheck the latest version of boehm_gc-7.6.4.recipe?

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,88 @@
+PATCHES="boehm_gc-$portVersion.patchset"
+
+ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
```
```diff
-ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
-SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
+SECONDARY_ARCHITECTURES="x86"
```

fbrosson replied:
I'm suggesting to mark boehm_gc as broken on  x86_gcc2 because it needs libatomic_ops which does not build on that arch.
That said, I'm not against keeping the `?x86_gcc2` for `ARCHITECTURES` only.

fbrosson commented:
Thanks @ivmai for the update!
Good news: I've just tested the recipe on both x86_64 and x86 secondary arch.
And on x86_64 I have also built and tested w3m against Boehm's libgc (and it runs perfectly.)
I'll test w3m and ecl on x86.

The only remaining issue is with `TEST()`, which fails with:
```
/sources/gc-7.6.4/tests/thread_leak_test.c:66: undefined reference to `GC_pthread_create'
/sources/gc-7.6.4/tests/thread_leak_test.c:79: undefined reference to `GC_pthread_join'
```
But I think it will be OK to merge even if we can't run the test suites yet.

ivmai commented:
> -ARCHITECTURES="?x86_gcc2 ?x86 x86_64"
> +ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
> -SECONDARY_ARCHITECTURES="?x86_gcc2 ?x86"
> +SECONDARY_ARCHITECTURES="x86"

Done.

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,88 @@
+SOURCE_DIR="gc-$portVersion"
+PATCHES="boehm_gc-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
```
```diff
-ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
```
Shall we mark x86 primary as tested?
I've just built boehm_gc and w3m on x86 primary arch and w3m is working like a charm.

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.0.recipe:
```diff
@@ -10,7 +10,7 @@ COPYRIGHT="1988, 1989 Hans-J. Boehm, Alan J. Demers
 	1999-2011 Hewlett-Packard Development Company, L.P."
-LICENSE="BOEHM"
+LICENSE="Boehm"
 REVISION="2"
```
**Question:** Should we increment `REVISION` (here, for the old recipe, `boehm_gc-7.6.0.recipe`), since we're making a cosmetic change in `LICENSE`?

(The rule is that if we are not sure we can keep `REVISION` unchanged, then we should increment it. But I don't think it would be worth incrementing it just to trigger a new build, which would be almost identical to the previous one.)

@mmlr and/or @pulkomandy, what would you recommend here?

mmlr replied:
The resulting package would be (even visibly) different, so the revision needs to be bumped.

alaviss commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,88 @@
+	"
+BUILD_PREREQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	cmd:autoconf
```
```diff
-	cmd:autoconf
+	cmd:autoreconf

alaviss commented on dev-libs/boehm_gc/boehm_gc-7.6.4.recipe:
```diff
@@ -0,0 +1,88 @@
+	autoconf -f
+	runConfigure ./configure --enable-large-config \
+		--enable-parallel-mark \
+		--disable-dependency-tracking
```
```diff
-	libtoolize -fci
-	autoconf -f
+	autoreconf -vfi
 	runConfigure ./configure --enable-large-config \
-		--enable-parallel-mark \
-		--disable-dependency-tracking
+		--enable-parallel-mark

ivmai commented:
Done (REVISION, autoreconf). Also I've added -Werror compiler option (to catch warnings if any) and reporting of errno on sem_wait fail.

@alaviss could you please test new recipe on x64 and run GC_PRINT_STATS=1 ./gctest

fbrosson commented on dev-libs/boehm_gc/boehm_gc-7.6.0.recipe:
```diff
@@ -10,8 +10,8 @@ COPYRIGHT="1988, 1989 Hans-J. Boehm, Alan J. Demers
-REVISION="2"
+LICENSE="Boehm"
+REVISION="3"
 SOURCE_URI="http://www.hboehm.info/gc/gc_source/gc-$portVersion.tar.gz"
```
Shall we use this variant?
```
SOURCE_URI="https://github.com/ivmai/bdwgc/releases/download/v$portVersion/gc-$portVersion.tar.gz
	http://www.hboehm.info/gc/gc_source/gc-$portVersion.tar.gz"
```
Advantages:
1) I uses https instead of http.
2) If GitHub is down, haikuporter will fallback to hboehm.info.

(Reminder: the 80 chars limit does not apply to `SOURCE_URI`.)

ivmai replied:
> SOURCE_URI="https://github.com/ivmai/bdwgc/releases/download/v$portVersion/gc-$portVersion.tar.gz
>	http://www.hboehm.info/gc/gc_source/gc-$portVersion.tar.gz"

Done

fbrosson commented:
Here're the last two blocks of output I get on x86_64 when I run `GC_PRINT_STATS=1 ./gctest`:
```
--> Marking for collection #77 after 938688 allocated bytes
Marked from 5 dirty pages
GC #77 freed 2768560 bytes, heap 7768 KiB
World-stopped marking took 1 msecs (0 in average)
In-use heap: 27% (2114 KiB pointers + 3 KiB other)
0 finalization entries; 0/0 short/long disappearing links alive
0 finalization-ready objects; 0/0 short/long links cleared
Finalize plus initiate sweep took 0 + 0 msecs

--> Marking for collection #78 after 1024224 allocated bytes
Marked from 3 dirty pages
GC #78 freed 676544 bytes, heap 7768 KiB
sem_wait failed: errcode= -2147483646sem_wait failed
Abort
```

fbrosson commented:
@ivmai, in case you would like to split the patchset in two or more parts, you can have haikuporter create/update the patchset for you. Here's how it works.
* Start with a clean work directory: `haikuporter -c boehm_gc`
* Apply the current patchset: `haikuporter -B boehm_gc`
* `cd work-7.6.4/sources/gc-7.6.4`
* Add your changes to any files, and use git to commit any changes you made inside any subdirectory of `work-7.6.4/sources/gc-7.6.4`. You can use git in here as you would on any project.
* `cd ../../..`
* Optional: `cp patches/boehm_gc-7.6.4.patchset patches/boehm_gc-7.6.4.patchset.bak`
* Generate the new patchset: `haikuporter -e boehm_gc`

This allows e.g. to split the patch into several parts, so you can add your changes with you as the author of your changes, and keep the previous authors for the changes they did, even if they required some synchronisation for the line offsets and/or adjustements with modified diff contexts.

ivmai commented:
> sem_wait failed: errcode= -2147483646

Could someone please look into sem_wait implementation in Haiku? What does the above errno code mean?

alaviss commented:
@ivmai It's `EACCES`.

fbrosson commented:
Hmm, I can't help you here...
All I might add is that `-2147483646 = 2 - 2^31`, but this is trivial and does not answer the question.

fbrosson commented:
```
error -2147483646
0x80000002: Permission denied
```

ivmai commented:
> It's EACCES.

> error -2147483646
> 0x80000002: Permission denied

A conforming implementation of sem_wait cannot have EACCES error. And what is the reason of such error? How should we handle it?

ivmai commented:
@alaviss do I need to change something in the recipe?

ivmai commented:
> BTW, if you fix all the tests it will even be possible to drop the previous recipe. So this would be great!

I will try. What's the preferred Haiku distributive? https://download.haiku-os.org/ ?

alaviss commented:
> @alaviss do I need to change something in the recipe?

It looks good enough to me, although you should drop `-Werror` before merge.

@fbrosson This library would be directly responsible for memory leaks for any application that uses it if there are any issues. I would prefer having the library tested, or having an issue filed about the test failures.

> A conforming implementation of sem_wait cannot have EACCES error.

It's [not required][0] that a conforming implementation to not have a specific error.

> It was the consensus of the standard developers that to allow the conformance document to state that an error occurs and under what conditions, but to disallow a statement that it never occurs, does not make sense. It could be implied by the current wording that this is allowed, but to reduce the possibility of future interpretation requests, it is better to make an explicit statement.

>  And what is the reason of such error? How would we handle it?

We would need @pulkomandy, @waddlesplash, @mmlr and @korli to help with this

[0]: http://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xsh_chap02.html#tag_22_02_03

ivmai commented:
> It looks good enough to me, although you should drop -Werror before merge.

Done

fbrosson commented:
> @fbrosson This library would be directly responsible for memory leaks for any application that uses it if there are any issues. I would prefer having the library tested, or having an issue filed about the test failures.

Yep, you are right. Let's wait until we can really test it.

BTW, with the current contents of this PR, I get these results:
* x86 second arch (hrev51795):
```
PASS: cordtest
./test-driver: line 107: 42313 Abort                   "$@" > $log_file 2>&1
FAIL: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
PASS: threadkey_test
FAIL: subthreadcreate_test
PASS: initsecondarythread_test
PASS: disclaim_test
PASS: disclaim_bench
============================================================================
Testsuite summary for gc 7.6.4
============================================================================
# TOTAL: 14
# PASS:  12
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0
```
* On x86_64 (hrev51727):
```
PASS: cordtest
./test-driver: line 107:  7269 Abort                   "$@" > $log_file 2>&1
FAIL: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
```
Then it KDLs. I can **es**cape the KDL, move the mouse, and even type any command e.g. uname, but it KDLs again when the command exits.
Well, I guess we really need to fix this before we merge.

ivmai commented:
> Some issues, mainly on x86_64.

mmlr commented:
Please open bug reports at dev.haiku-os.org, definitely for the KDL and possibly for the semaphore issue.

I only took a cursory look at the sem issue. A sem_wait implies a POSIX realtime sem, of which the implementation only directly returns EACCES on creation and deletion. It does build upon native semaphores, which should only yield permission denial if a user thread tried to acquire a kernel sem. These should be logged to the syslog, so a look there may give some more hints.

ivmai commented:
1. The tests should pass now. Please check.
2. I haven't filed the bug reports yet (besides EACCES another bug in Haiku is pthread_join which returns ESRCH in case of many threads were created).

fbrosson commented:
Thanks @ivmai for the update. Things look better :)

* Excellent news on x86 (primary as well as secondary), because the 2 tests that were previously failing, `cordtest` & `subthreadcreate_test`, are now OK:
```
PASS: cordtest
PASS: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
PASS: threadkey_test
PASS: subthreadcreate_test
PASS: initsecondarythread_test
PASS: disclaim_test
PASS: disclaim_bench
============================================================================
Testsuite summary for gc 7.6.4
============================================================================
# TOTAL: 14
# PASS:  14
```
* On x86_64 `gctest` was also previously failing, and now it is OK. But the crash still happens after `threadleaktest`, and the system becomes unstable after that test:
```
PASS: cordtest
PASS: gctest
PASS: leaktest
PASS: middletest
PASS: smashtest
PASS: hugetest
PASS: realloc_test
PASS: staticrootstest
PASS: threadleaktest
```

fbrosson commented:
@alaviss, on x86_64 it's better than before, but the system becomes unstable. Well, this happens on my system.

extrowerk commented:
Any news here? It is required for guile... :smile:

fbrosson commented:
@ivmai, @alaviss, @extrowerk, I did not merge this PR but I'm OK with a merge.

Maybe what caused my system to KDL was not the shared libs but the test suite.
So please feel free to merge this PR if you wish.
(I'm not doing it myself because I would not know how to handle the topic if things went wrong.)

alaviss commented:
@fbrosson I think we should mark x86_64 as broken. A reference to the filed Trac ticket should be included as well.

fbrosson commented:
@ivmai, I can push a tiny commit (to this PR) with the change suggested by @alaviss if you agree, and then merge.
Shall I do so?

Regarding the references for possible Trac tickets, I don't have any.

alaviss commented:
@fbrosson Can you please reproduce the KDL and submit it to Trac? I'm unable to run any Haiku VM now due to some problems :(

ivmai commented:
> I can push a tiny commit (to this PR) with the change suggested by @alaviss if you agree, and then merge. Shall I do so?
@fbrosson Yes, please. (I don't think there's a regression since 7.6.0 but the tests are more aggressive in 7.6.4)

fbrosson commented:
Sorry for this late reply. (I was AFK.)

@alaviss: I'll try to reproduce the KDL, but I need to update my system because I'm using a not so recent hrev.

@ivmai: Cool. I'll mark `x86_64` as broken. Then we can merge if everyone is OK with that.

@extrowerk: Are you running x86_gcc2 or x86_64? If it's x86_64, do the tests KDL your system?

fbrosson commented:
I've prepared a possible commit message:
```
boehm_gc: add recipe for version 7.6.4.

* Rename license file (BOEHM -> Boehm)
* Remove --enable-thread-local-alloc
* Sort PROVIDES[_devel], BUILD_PREREQUIRES.
* Drop the libtool files after make install
* Add debug-info package
* Remove CFLAGS+=-fPIC for make check
* Remove --enable-threads, --disable-debug configure options
* Use autoreconf instead of libtoolize+autoconf
* Add fallback SOURCE_URI (referring to hboehm.info)
* Workaround sem_wait and pthread_join fails
* Mark 7.6.4 as broken on x86_64 for now.
```
It's a compilation of the commit messages in this PR.
(Well, I have dropped a couple of not very important lines.)

What do you think? Suggestions welcome ;)
Shall I merge using the suggested commit message (or an improved commit message)?

extrowerk commented:
@fbrosson : on x86_64 all test passed, no kdl yet, tested 3 times already.

ivmai commented:
> I've prepared a possible commit message:

Looks good to me.

fbrosson commented:
@extrowerk

> on x86_64 all test passed, no kdl yet, tested 3 times already.

This is great news! It looks like the problem was only my system. Thanks @extrowerk!

@ivmai, @extrowerk, @alaviss: So, after all, shall we merge with x86_64 marked as tested?
(I think this is OK.)
Let me eat the last commit I added a couple of hours ago. (I'll drop the last line of the commit message, of course.)

ivmai commented:
OK to merge.

fbrosson commented:
Thanks @ivmai for your PR and sorry for the delay for the merge.
(I should have updated my system...)

ivmai commented:
Thank you all for the reviews.
